### DISCUSSION TITLE
Confused by Glib's `__attribute__(cleanup)` and MSVC compiling

### DISCUSSION THREAD

**Discussion by felixf4xu** [microsoft/vcpkg#38446] [General]
Hi,

I tried to port a GObject/Glib based library to vcpkg, but in its meson.build file it disables msvc compiler. The reason is that Glib is using `g_autoptr` which is using `__attribute__(cleanup)` and it's not available in msvc. I understand that, so far.

Then I tried to compile Glib on Windows locally, it compiled with MSVC and it succeeded. 

So I was confused. If Glib can't be compiled by MSVC, why in vcpkg default compiling triplet, it succeeded?

**Comment by dg0yt** [comment]
There is a glib port, and it builds for windows (MSVC) tripelts. And also the glib README indicates that MSVC works, https://github.com/winlibs/glib/blob/master/README.win32. This is what we know.

But we have zero knowledge about this "GObject/Glib based library" you are trying to port. Maybe it is only this software, but not glib itself, which is using `g_autoptr`? This macro is indeed documented to be "only supported on GCC and clang. This macro is not defined on other compilers and should not be used in programs that are intended to be portable to those compilers.", https://docs.gtk.org/glib/auto-cleanup.html. But this doesn't imply that `glib` is unsupported on MSVC.

**Comment by felixf4xu** [comment]
Thanks for the comment, now I understand that vcpkg can build glib itself successfully, because glib only defines `g_autoptr` but never uses it.

Then I checked GTK source code, I see
>* GTK+ types now support the GLib g_autoptr() facility for
  automatic cleanup of automatic variables. GTK+ itself does
  **not** use g_autoptr() internally.

So I also understood that GTK can be built by vcpkg (msvc).

As to the lib I mentioned which depends on glib, it unfortunately uses g_autoptr, so it can't be built by vcpkg(msvc)
